### PR TITLE
feat: endpoints2: endpoint provider middleware and final integration

### PR DESCRIPTION
**Pull request by lucix-aws**
## Description of changes
Implement middleware and endpoint parameter binding generators to call endpoint providers per-operation at runtime.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: runtime/protocol/http/common/src/aws/smithy/kotlin/runtime/http/endpoints/EndpointResolver.kt [hunk 1]
```diff
@@ -5,9 +5,15 @@
...
+
 /**
- * Resolves endpoints for a given service client
+ * Resolves endpoints for a given service client.
+ *
+ * This is a static version of the otherwise-generated endpoint resolver interface and is intended for use by internal
+ * clients within the runtime.
  */
+@InternalApi
 public fun interface EndpointResolver {
```
Comment (ianbotsf) [inline runtime/protocol/http/common/src/aws/smithy/kotlin/runtime/http/endpoints/EndpointResolver.kt:17]: **Question**: Is there no supported way to specify a custom endpoint resolver anymore?

File: smithy-kotlin-codegen/src/main/kotlin/software/amazon/smithy/kotlin/codegen/rendering/endpoints/EndpointParametersGenerator.kt [hunk 2]
```diff
@@ -38,6 +38,8 @@ class EndpointParametersGenerator(
                 name = CLASS_NAME
                 namespace = "${settings.pkg.name}.endpoints"
             }
+
+        fun getParamKotlinName(param: Parameter): String = param.name.toKotlin()
     }
 
     private val params: List<KotlinEndpointParameter> = rules.parameters.toList()
```
Comment (aajtodd) [inline smithy-kotlin-codegen/src/main/kotlin/software/amazon/smithy/kotlin/codegen/rendering/endpoints/EndpointParametersGenerator.kt:45]: GH won't let me comment on the unchanged line but we generate a companion object for `EndpointParameters` like this:

```kotlin
    public companion object {
        public operator fun invoke(block: Builder.() -> Unit): EndpointParameters = Builder().apply(block).build()
    }
```

This function should probably be marked `inline` to avoid unnecessary backing lambdas everywhere we have to construct endpoint parameters in generated code (contributes to bloated artifact size)

File: smithy-kotlin-codegen/src/main/kotlin/software/amazon/smithy/kotlin/codegen/rendering/protocol/HttpBindingProtocolGenerator.kt [hunk 3]
```diff
@@ -293,7 +293,7 @@ abstract class HttpBindingProtocolGenerator : ProtocolGenerator {
         val pathBindings = requestBindings.filter { it.location == HttpBinding.Location.LABEL }
 
         if (pathBindings.isNotEmpty()) {
-            writer.openBlock("val pathSegments = listOf(", ")") {
+            writer.openBlock("val pathSegments = listOf<String>(", ")") {
```
Comment (ianbotsf) [inline smithy-kotlin-codegen/src/main/kotlin/software/amazon/smithy/kotlin/codegen/rendering/protocol/HttpBindingProtocolGenerator.kt:296]: **Question**: Why was this change necessary?

File: smithy-kotlin-codegen/src/main/kotlin/software/amazon/smithy/kotlin/codegen/rendering/endpoints/ResolveEndpointMiddlewareGenerator.kt [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+import software.amazon.smithy.rulesengine.traits.ClientContextParamsTrait
+import software.amazon.smithy.rulesengine.traits.ContextParamTrait
+import software.amazon.smithy.rulesengine.traits.StaticContextParamsTrait
+
+/**
+ * Generates resolver middleware for a specific endpoint provider signature.
+ * The binding of parameter values is generated separately in [EndpointParameterBindingGenerator], and that generated
+ * implementation is plugged in here.
+ */
+class ResolveEndpointMiddlewareGenerator(
```
Comment (aajtodd) [inline smithy-kotlin-codegen/src/main/kotlin/software/amazon/smithy/kotlin/codegen/rendering/endpoints/ResolveEndpointMiddlewareGenerator.kt:31]: Rather than comment on each individual location I'm going to make one long comment on the resulting generated code.

Taken from S3 it seems we are generating the following middleware and middleware registration:

Current middleware being generated (per service):
```kotlin
public class ResolveEndpointMiddleware<I, O>(
    private val endpointProvider: EndpointProvider,
    private val buildParams: EndpointParameters.Builder.(input: I) -> Unit,
) : InlineMiddleware<I, O> {
    private lateinit var endpoint: Endpoint

    public override fun install(op: SdkHttpOperation<I, O>) {
        op.execution.initialize.intercept { req, next ->
            val params = EndpointParameters { buildParams(req.subject) }
            endpoint = endpointProvider.resolveEndpoint(params)
            next.call(req)
        }

        op.execution.mutate.intercept { req, next ->
            setResolvedEndpoint(req, endpoint)
            next.call(req)
        }
    }
}
```

Example current operation setup/registration:
```kotlin
    override suspend fun abortMultipartUpload(input: AbortMultipartUploadRequest): AbortMultipartUploadResponse {
        ...

        op.install(ResolveEndpointMiddleware<AbortMultipartUploadRequest, AbortMultipartUploadResponse>(config.endpointProvider) { input ->
            bindAwsBuiltins(config)
            bucket = input.bucket
        })
      
        ...
```

-----------------

Couple things stand out to me right away.

The middleware takes a `buildParams` lambda which is passed `req.subject` to bind the input shape members to the endpoint parameters (if there are any). This isn't really necessary though as we have access to the operation input when creating the middleware (and is what you are using already anyway, e.g. `input.bucket` rather than the parameter from the lambda). This relationship is a bit confusing and seems like it's there only because we want to do it in middleware.

I'm almost wondering (and this is just an idea I'm throwing out there I don't even know if I like it yet) if we shouldn't just build the endpoint parameters in the operation setup before middleware registration:

```kotlin
val endpointParameters = EndpointParameters {
    bindBuiltins(config)
    bucket = input.bucket
    ...
}



op.install(ResolveEndpointMiddleware(config.endpointProvider, endpointParameters))

```

`ResolveEndpointMiddleware` could then just be a normal `ModifyRequestMiddleware`

--------

We could update the `EndpointResolver` and `ResolveEndpoint` middleware types to be re-usable such that we don't have to generate a separate resolver type per/client but instead use generics:

something like:
```kotlin
public class ResolveEndpoint<T> {
    private val resolver: EndpointResolver<T>,
    private val parameters: T
): ModiifyRequestMiddleware {
    override suspend fun modifyRequest(req: SdkHttpRequest): SdkHttpRequest {
        val endpoint = resolver.resolve(parameters)
        setResolvedEndpoint(req, endpoint)
        ...
    }   
}
```

Resolver w/generics
```kotlin
public fun EndpointResolver<T> {
    public suspend fun resolve(parameters: T): Endpoint
}
```

The generated client config would then be:

```kotlin
public class Config ... {
     ...
     // specialized to whatever the EndpointParameters type is that IS generated per/service
     public val endpointProvider: EndpointProvider<EndpointParameters> = builder.endpointProvider ?: DefaultEndpointProvider()

}
```

Whether we do this with generics or keep per/service generated resolver type isn't super important it's just one idea to try and re-use as much as possible and not need custom codegen for AWS vs generic smithy models.

File: smithy-kotlin-codegen/src/main/kotlin/software/amazon/smithy/kotlin/codegen/rendering/protocol/HttpBindingProtocolGenerator.kt (hunk 3 above)
Comment (lucix-aws) [reply smithy-kotlin-codegen/src/main/kotlin/software/amazon/smithy/kotlin/codegen/rendering/protocol/HttpBindingProtocolGenerator.kt:296]: So this is a side effect of the bucket path preprocessing that now happens on the s3 model.

If you remove both the path segment and the httpLabel trait, our serializer generator renders the bucket to be serialized as part of the payload, which is wrong.

Since the path serializer references off of the httpTrait path itself, I just left the httpLabel trait alone and when we go to render the path serializer it simply ignores the orphaned httpLabel members.

A side effect of that is the list of paths generated here can now be empty, hence the need to explicitly add the String to the declaration.

There may be a better way to express "do not serialize this" in smithy that I'm not aware of.

File: smithy-kotlin-codegen/src/main/kotlin/software/amazon/smithy/kotlin/codegen/rendering/endpoints/ResolveEndpointMiddlewareGenerator.kt [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+    }
+
+    /**
+     * Render the assignment of an endpoint param value within a builder. The SEP dictates a specific source order to
+     * use when determining how to bind:
+     * 1. staticContextParams (from operation shape)
+     * 2. contextParam (from member of operation input shape)
+     * 3. clientContextParams (from service shape)
+     * 4. builtin binding
+     * 5. builtin default
```
Comment (aajtodd) [inline smithy-kotlin-codegen/src/main/kotlin/software/amazon/smithy/kotlin/codegen/rendering/endpoints/ResolveEndpointMiddlewareGenerator.kt:125]: Do we have any instances of (5)? I'm assuming `bindAwsBuiltins` is type (4) and why we render it _before_ delegating to `EndpointParameterBindingGenerator` in `aws-sdk-kotlin`?

Comment (lucix-aws) [reply smithy-kotlin-codegen/src/main/kotlin/software/amazon/smithy/kotlin/codegen/rendering/endpoints/ResolveEndpointMiddlewareGenerator.kt:125]: Correct, and (5) is essentially fulfilled by the config field defaults.

File: runtime/protocol/http/common/src/aws/smithy/kotlin/runtime/http/endpoints/EndpointProvider.kt [hunk 6]
```diff
@@ -5,19 +5,14 @@
...
-
 /**
  * Resolves endpoints for a given service client.
- *
- * This is a static version of the otherwise-generated endpoint resolver interface and is intended for use by internal
- * clients within the runtime.
  */
-@InternalApi
-public fun interface EndpointResolver {
+public fun interface EndpointProvider<T> {
```
Comment (ianbotsf) [inline runtime/protocol/http/common/src/aws/smithy/kotlin/runtime/http/endpoints/EndpointProvider.kt:11]: **Question**: Is there a semantic difference between a "provider" and a "resolver"? This interface's name changed but its KDoc and single method name still refer to "resolution" rather than "providing".

File: smithy-kotlin-codegen/src/main/kotlin/software/amazon/smithy/kotlin/codegen/model/RulesEngineExt.kt [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+ * Get the symbol for an endpoint parameter type. All endpoint parameter members are nullable.
+ */
+fun ParameterType.toSymbol(): Symbol =
+    when (this) {
+        ParameterType.STRING -> KotlinTypes.String
+        ParameterType.BOOLEAN -> KotlinTypes.Boolean
+    }
+        .toBuilder()
+        .boxed()
+        .build()
```
Comment (ianbotsf) [inline smithy-kotlin-codegen/src/main/kotlin/software/amazon/smithy/kotlin/codegen/model/RulesEngineExt.kt:36]: **Comment**: `.toBuilder().boxed().build()` is a thing we do in a few other places now. Feels like this is officially convenience-method-ready.

File: smithy-kotlin-codegen/src/main/kotlin/software/amazon/smithy/kotlin/codegen/model/RulesEngineExt.kt [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+
+/**
+ * Get the writable literal for a rules engine value.
+ */
+fun Value.toLiteral(): String =
+    when (this) {
+        is Value.String -> expectString().doubleQuote()
+        is Value.Bool -> if (expectBool()) "true" else "false"
```
Comment (ianbotsf) [inline smithy-kotlin-codegen/src/main/kotlin/software/amazon/smithy/kotlin/codegen/model/RulesEngineExt.kt:44]: **Nit**: `expectBool().toString()`

File: smithy-kotlin-codegen/src/main/kotlin/software/amazon/smithy/kotlin/codegen/rendering/endpoints/ResolveEndpointMiddlewareGenerator.kt [hunk 9]
```diff
@@ -31,6 +31,7 @@ import software.amazon.smithy.rulesengine.traits.StaticContextParamsTrait
 class ResolveEndpointMiddlewareGenerator(
     private val ctx: ProtocolGenerator.GenerationContext,
     private val writer: KotlinWriter,
+    private val renderPostResolution: () -> Unit = {},
 ) {
```
Comment (ianbotsf) [inline smithy-kotlin-codegen/src/main/kotlin/software/amazon/smithy/kotlin/codegen/rendering/endpoints/ResolveEndpointMiddlewareGenerator.kt:35]: **Question**: Where is a non-default argument passed for `renderPostResolution`? I couldn't find anywhere in this PR or the **aws-sdk-kotlin** one...

**Review by ianbotsf — APPROVED** [review]
Nice work! Nothing blocking, just questions and low-priority suggestions.

File: runtime/protocol/http/common/src/aws/smithy/kotlin/runtime/http/endpoints/EndpointProvider.kt (hunk 6 above)
Comment (aajtodd) [reply runtime/protocol/http/common/src/aws/smithy/kotlin/runtime/http/endpoints/EndpointProvider.kt:11]: I don't think there is, I preferred resolver with a single `resolve` method but this matches the SEP/ref arch terminology I believe.

File: smithy-kotlin-codegen/src/main/kotlin/software/amazon/smithy/kotlin/codegen/rendering/endpoints/ResolveEndpointMiddlewareGenerator.kt [hunk 10]
```diff
@@ -0,0 +1,153 @@
...
+        writer.closeAndOpenBlock(") : #T<I, O> {", RuntimeTypes.Http.Operation.InlineMiddleware)
+        renderClassMembers()
+        writer.write("")
+        renderInstall()
+        writer.closeBlock("}")
+    }
+
+    private fun renderConstructorParams() {
+        writer.write("private val endpointProvider: #T,", EndpointProviderGenerator.getSymbol(ctx.settings))
+        writer.write("private val buildParams: #T.Builder.(input: I) -> Unit,", EndpointParametersGenerator.getSymbol(ctx.settings))
```
Comment (aajtodd) [inline smithy-kotlin-codegen/src/main/kotlin/software/amazon/smithy/kotlin/codegen/rendering/endpoints/ResolveEndpointMiddlewareGenerator.kt:58]: IDK why this irks me so much but it does. I don't have a better suggestion though so it's fine for now.


Part of my distaste for this is that it results in backing lambdas for each operation, e.g.

```
> javap DefaultS3Client\$deleteBucket\$2                  
Warning: File ./DefaultS3Client$deleteBucket$2.class does not contain class DefaultS3Client$deleteBucket$2
Compiled from "DefaultS3Client.kt"
final class aws.sdk.kotlin.services.s3.DefaultS3Client$deleteBucket$2 extends kotlin.jvm.internal.Lambda implements kotlin.jvm.functions.Function2<aws.sdk.kotlin.services.s3.endpoints.EndpointParameters$Builder, aws.sdk.kotlin.services.s3.model.DeleteBucketRequest, kotlin.Unit> {
  final aws.sdk.kotlin.services.s3.DefaultS3Client this$0;
  aws.sdk.kotlin.services.s3.DefaultS3Client$deleteBucket$2(aws.sdk.kotlin.services.s3.DefaultS3Client);
  public final void invoke(aws.sdk.kotlin.services.s3.endpoints.EndpointParameters$Builder, aws.sdk.kotlin.services.s3.model.DeleteBucketRequest);
  public java.lang.Object invoke(java.lang.Object, java.lang.Object);
}
```

The alternative isn't much better though as it would require us to generate a middleware per/operation anyway

**Review by aajtodd — APPROVED** [review]

File: smithy-kotlin-codegen/src/main/kotlin/software/amazon/smithy/kotlin/codegen/rendering/endpoints/ResolveEndpointMiddlewareGenerator.kt (hunk 9 above)
Comment (lucix-aws) [reply smithy-kotlin-codegen/src/main/kotlin/software/amazon/smithy/kotlin/codegen/rendering/endpoints/ResolveEndpointMiddlewareGenerator.kt:35]: The default implementation on the `ProtocolGenerator` method does

File: smithy-kotlin-codegen/src/main/kotlin/software/amazon/smithy/kotlin/codegen/rendering/endpoints/ResolveEndpointMiddlewareGenerator.kt (hunk 10 above)
Comment (lucix-aws) [reply smithy-kotlin-codegen/src/main/kotlin/software/amazon/smithy/kotlin/codegen/rendering/endpoints/ResolveEndpointMiddlewareGenerator.kt:58]: Yeah, I had the per-operation middleware before refactoring to this, which was much worse IMO.
